client/client.go

```diff
@@ -188,3 +195,133 @@ func Ping(address string, config *Config) (bool, time.Duration) {
+	defer refclient.Reset()
+	ds := grpcurl.DescriptorSourceFromServer(ctx, refclient)
+
+	// if there's a verb (e.g. list)
+	if len(grpcConfig.Verb) > 0  {
+		switch verb := grpcConfig.Verb; verb {
+		case "list":
+			return GRPCList(ds, client, grpcConfig.Service)
+		default:
+			// ValidateAndSetDefaults from core.endpoint should prevent
+			// entering here
+			return false, nil, fmt.Errorf("%s not implemented", verb)
+		}
+		// Could extend here to implement health check protocol, as
+		// in https://github.com/TwiN/gatus/issues/157 or 'describe'
+		// like in grpcurl.
+	} else {
+		// RPC
+		return GRPCInvokeRPC(client, ctx, ds, body, grpcConfig.Service)
+	}
+}
+
+// List the services of a server or service
+// Returns:
+// - bool: whether the connection was stablished
+// - []byte: byte representation of services separated by "\n"
+// - error: if there was an error
+func GRPCList(ds grpcurl.DescriptorSource, client *grpc.ClientConn, service string) (bool, []byte, error) {
+	if len(service) == 0 {
+		services, err := grpcurl.ListServices(ds)
+		if err != nil {
+			// Unsure if this means succesfull connection or not
+			return true, nil, fmt.Errorf("Error listing services of server: %w", err)
+		}
+		if len(services) == 0 {
+			return true, nil, nil
+		} else {
+			serviceList := []byte(strings.Join(services, "\n"))
+			return true, serviceList, nil
+		}
+	} else {
+		methods, err := grpcurl.ListMethods(ds, service)
+		if err != nil {
+			return false, nil, fmt.Errorf("Error listing methods of service: %w", err)
+		}
+		if len(methods) == 0 {
+			return true, nil, nil
+		} else {
+			methodList := []byte(strings.Join(methods, "\n"))
+			return true, methodList, nil
+		}
+	}
+}
+
+// Invoke an RPC via gRPC
+// Returns:
+// - bool: whether the connection was stablished
+// - []byte: byte representation of services separated by "\n"
+// - error: if there was an error
+func GRPCInvokeRPC(client *grpc.ClientConn, ctx context.Context, ds grpcurl.DescriptorSource, body string, service string) (bool, []byte, error) {
+	options := grpcurl.FormatOptions{
+		EmitJSONDefaultFields: true,
+		AllowUnknownFields:    true,
+		IncludeTextSeparator:  true,
+	}
+
+	// Data to send to server
+	data := strings.NewReader(body)
+
+	rf, formatter, err := grpcurl.RequestParserAndFormatter(grpcurl.Format("json"), ds, data, options)
+	if err != nil {
+		return false, nil, fmt.Errorf("Error: %w", err)
+	}
+
+	buffer := new(strings.Builder)
+	handler := &grpcurl.DefaultEventHandler{
+		Out:            buffer,
+		Formatter:      formatter,
+		VerbosityLevel: 0, // Zero is the default. Increasing this value
+		                   // increases the verbosity of the output of
+		                   // the RPC invocation, adding request
+		                   // destination, response header, etc. which
+		                   // messes up parsing the "BODY".
+	}
+
+	var headers []string
+	err = grpcurl.InvokeRPC(ctx, ds, client, service, headers, handler, rf.Next)
```

Comment:
Something is missing here. Gatus is erroring out in our staging gateway with `server does not support the reflection API` but with grpcurl I can make the same request and even `list` and see the `grpc.reflection.v1alpha.ServerReflection`...

What could it be?

Reply:
It works for other servers though.

Reply:
Or maybe the issue is not on this code?

Reply:
Can you share which server is causing the issue